pipeline/matching/generators/clustering.py

```diff
@@ -0,0 +1,86 @@
+import math
+from typing import Iterator
+
+from pipeline.matching.core import MatchGenerator
+from pipeline.types import Match, MatchingInput, MatchMetadata
+
+GENERATOR_CLUSTERING = "clusteringGenerator"
+
+# TODO: Add type annotations to auxilary functions
+def cosine_similarity(v1, v2):
+    "compute cosine similarity of v1 to v2: (v1 dot v2)/{||v1||*||v2||)"
+    sumxx, sumxy, sumyy = 0, 0, 0
+    for i in range(len(v1)):
+        x = v1[i]
+        y = v2[i]
+        sumxx += x * x
+        sumyy += y * y
+        sumxy += x * y
+    if sumxx * sumyy == 0:
+        return 0
+    return sumxy / math.sqrt(sumxx * sumyy)
+
+
+def closest_match(user1, users):
+    """
+    Summary:
+        loop over all other users to find match with closest aka max cosine similarity --
+        predefined function "cosine_similarity(user1[1],user2[1])"
+        think of users as an 11 dimensional point and trying to find the closest distance between two points
+    Performance: O(?)
+    """
+    cos_sim = 0
+    user1_match = ""
+    for user in users:
+        if user[0] == user1[0]:
+            continue
+        if cosine_similarity(user[1], user1[1]) > cos_sim:
+            cos_sim = cosine_similarity(user[1], user1[1])
```

Comment:
Can you think of a way to make these 2 lines more efficient since we are making 2 calls to cosine_similarity?